### PR TITLE
Rewrite dependency extraction to use mod list system V2

### PR DESCRIPTION
**Pull request by LexManos**
This is the successor to #4081, this is being opened mainly for comments from people.
1.13 will require mods to supply unique maven style identifiers. This is not up for debate its something we have NEEDED for years,.
For 1.12 I would like to introduce support for this, with some backwards compatibility things in mind.
So ya, i'm not great at explaining things so lets get down to the dirty:

**Repositories**:
We will now attempt to determine the 'normal' libraries folder. This is done by trying to find the jar that one of our apache libraries is in and then seeing if it's in the 'maven style' folder structure that we expect. If it is not, we will just use ./libraries/ as the directory. This is overrideable via the forge.lib_folder system variable.
  
 ALL repos will be asked for an artifact when resolving. First artifact found for the exact version wins. Except for *-SNAPSHOT artifacts which will use the latest timestamp. More on that below.

**Mod Lists**:
The modlist json format is still the same as it has always been. I have plans to change it from Json to a plaintext format to hopefully ease the end user's fear of editing it. However this is on indefinite hold until I can get more feedback/input from people.
The Format I WISH to have would be something along the lines of:
>#This is a Comment
>Repository: libraries # This will use Java's Path API to determine if its absolute, or if its relative it will be relative to the mc home directory.
>Parent: parent_list.txt # Path works the same as Repository, but this points to another list file that will be read
>group:artifact:version:classifier@extention # Standard maven artifact format. Classifer and extension are optional. 

Things to note, it will parse every line after stripping comments. Which is why the example has comments at the end of the line.
And anything that is not starting with reserved keys {'Parent:' and 'Repository:'} will be treated as a artifact.
This format change is purely for end user ease of use. Json has syntax that could confuse the end user. 
It also doesn't support writing comments, this will. So that if we disable mods in the game, Forge can re-write the list, keeping the artifact in it, but commented out.
But ya.. As I said that change is on hold until I can get feedback.

**Contained Deps**:
This is the main reason for this PR. To make the library management of Forge a bit more standerdized and sane.
Before any coremods or normal mods are scanned Forge will now process all files in the /mods/ & /mods/1.12.2/ folders and any mod list files passed in via --modlist, /mods/mod_list.json and /mods/1.12.2/mod_list.json
For each jar it will check its Manifest for a ContainedDeps entry. This is the same entry that exists today. This is a space seperated list of files contained inside the jar that should be extracted. 
To promote standard clean folder structure, it will check for the entry in /META-INF/libraries/{entry} However for backwards compatibility, it will also just check for the full entry name without any prefix. I want to remove that compatibility in 1.13.

The fml.skipContainedDeps system variable is still implemented and respected in this process.
To determine where to put the extracted library, we will attempt to load the manifest for the contained jar. To solve the issue of 3rd party libraries not supporting the entries needed. External manifest files are also supported. Located at the same path as the contained dep just with an extra .meta tacked on to the end. 
We check the manifest for the 'Maven-Artifact' entry. If found, it will extract the file to the maven-style repository specified by the modlist the entry was in. Defaulting to the libraries repo found above.
For any artifact with versions ending in -SNAPSHOT (case insensitive) we will also check the manifest for a 'Timestamp' entry. Which is an integer since the epoch (standard timing format). It will keep a list of all snapshot files identified by that timestamp. 
If Maven-Artifact is NOT found, for backwards compatibility reasons (So, being removed in 1.13) the files will be extracted to the /mods/ directory. 

**Auto-Archiving**:
Now this is a controversial part of this, which is why it is disabled by default. But I have implemented a system that would automatically move any jar found in the mods folders to the libraries folder. If they have the Maven-Artifact manifest entry.
Adding it's artifact entry to the mod list. Basically, this would keep the 'toss mod into mods folder' style of installing mods that we have today. However it would change the removing of mods from 'delete the file in the mods folder' to 'delete the line in the modlist file'. I want to do this for many reasons, but the top two are: 1) Gets people used to using/editing the mods list. 2) It would put a spotlight on modders who do not follow the Maven-Artifact standard. Because they would be the only files left in the mods folder. But ya, this is disabled by default, and to do that I had to do some major hacks and created a virtual repository system. It just feels dirty writing it.

**Dependency Resolution:**
A potential benefit/extension of this system is to allow modders to specify dependencies in metadata. And then let Forge detect the correct versions of mods to run. Or libraries to load. Or error with nice error messages if there are conflicting dependencies. Basically an extension of our current dependency/sorting system we use for normal mods. Except it'd work for Coremods, and non-mod libraries. This part is not implemented because we need to discuss it and decide if people want it. And if they do what format it will take. But this is one of the powers of enforcing standards on identifying mods. So ya. a peek of the future?

**TODO:**
So there are things that arnt implemented yet, but those are not offically part of this PR. They are more ehre for talking points.
The only real TODO is to have @PaleoCrafter come along and build me up some unit tests to make things nice and robust.
Also we can discuss/give examples if people would like to talk.

**Warning:**
This thread needs be kept civil and on topic. Comments about the Auto-Arching, and Dependancy management need to be constructive. Not "don't do it!" I've already had enough of that discussion. There is a reason it's disabled by default for now.

### CODE REVIEW

**Comment by peterix** [comment]
We discussed that auto archiving point quite a bit, but I think we missed the 'put a spotlight on modders who do things wrong' angle.

Hinting at the problem or involving the users just doesn't seem like the direct and most effective approach.

Is it possible to detect that the mod is missing the `Maven-Artifact` entry in the dev environment, or that it is not correct?
If so, you could make missing this a fatal error in both 1.12 and 1.13. Only mod authors would ever see this sort of error, and it could point them to documentation (with why and how to fix it). 

But that doesn't expose people to just how much mess releasing two versions with the same version number makes...

So there's really this usability and user habits vs. teaching modders basics of versioning issue...

Can we possibly get someone from Curse to stop accepting/approving mods that do not do this correctly if they are for 1.12 and later?

**Comment by Niko-sk2x** [comment]
> If so, you could make missing this a fatal error in both 1.12 and 1.13. Only mod authors would ever see this sort of error, and it could point them to documentation (with why and how to fix it).

I think this could probably be only done in ForgeGradle, not in forge, as dev environment is different enough that there is unlikely to even be MANIFEST.MF for the information to be in. But this would also be a breaking change in ForgeGradle so probably would require a version bump. And is it actually a good idea for it to fail in that case?

**Comment by LexManos** [comment]
Merging to get this rolling and see how it goes. Was hoping to have unit tests but Paleo's been gone.
 We can continue the discussion on the other major changes. But as it sits to prevent the community from exploding this *should* be fully backwards compatible {except for internal FML API} and still function as the 'drop in mods dir and work' mentality.

**Comment by josephcsible** [comment]
This doesn't appear to be backwards compatible, at least in the Twitch launcher. On Forge 2655 it loads all of my mods correctly, but on 2656 it doesn't load any of them.

File: src/main/java/net/minecraftforge/fml/relauncher/libraries/LibraryManager.java [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+                        }
+                    }
+                }
+                else
+                {
+                    try
+                    {
+                        String timestamp = meta.getValue(TIMESTAMP);
+                        if (timestamp != null)
+                            timestamp = SnapshotJson.TIMESTAMP.format(new Date(Integer.parseInt(timestamp)));
```
Comment (JamiesWhiteShirt) [inline src/main/java/net/minecraftforge/fml/relauncher/libraries/LibraryManager.java:323]: ~~This limits the timestamp to Jan 25 21:31:23 1970, as this Date constructor takes a long in milliseconds. Did you mean `Long.parseLong(timestamp)`?~~

**Pull request merged**